Add mxid to signed event data (#37)
Informo/specs: content/information-distribution/signature.md

|    Field     |      Type      |                                    Description                                     |
|--------------|----------------|------------------------------------------------------------------------------------|
| `algorithm`  | `string`       | Algorithm used for signing the content. 🔧 Define allowed algorithms               |
| `sender_key` | `string`       | The public part of the key used for signing the event                              |
| `device_id`  | `string`       | 🔧 Optional, ID of the sending device, may be used for Megolm                      |
| `session_id` | `string`       | 🔧 Optional, ID of the session used to encrypt the message, may be used for Megolm |
| `signature`  | `string`       | Generated signature                                                                |
| `signed`     | `SignedObject` | Object to be signed                                                                |

Where:

* `SignedObject` is a map using the following structure:

| Parameter          |       Type       | Req. |              Description               |
|:-------------------|:-----------------|:----:|:---------------------------------------|

> | `sender`           | `string`         |  x   | The matrix user ID of the sender.      |

Reviewer: `sender` can be a bit ambiguous, as it could refer to the sender of the signed event. I'd rename it into `signatory`.

Author: The idea was to copy the matrix event structure, as it is meant to have the same value, but signatory is also a good choice as it's clear it is the user signing the event

> | `sender`           | `string`         |  x   | The matrix user ID of the sender.      |

Reviewer: ```suggestion
| `sender`           | `string`         |  x   | The matrix user ID of the sender. **Must** match the Matrix ID of the user that published the signature.      |
```

Author: I don't think this is really useful, since it is explained below

| `content`          | `object`         |  x   | The content of the event to be signed. |

> If the value of the `signed.sender` field does not match the original matrix

Reviewer: I find this sentence also ambiguous about which event is being talked about. Also, please pay attention to capitalising the first letter of names such as Matrix.

> | `signatory`        | `string`         |  x   | The matrix user ID of the sender. **Must** match the Matrix ID of the user that published the signature. |

Reviewer: ```suggestion
| `signatory`        | `string`         |  x   | The matrix user ID of the sender. **Must** match the Matrix ID of the user that generated the signature. |
```

Sorry, bad thinking on my side.

> event sender, the signed event **must** be considered as badly signed.

Reviewer: ```suggestion
event sender, the signature **must** be considered as invalid.
```

`m.room.encrypted`
events](https://matrix.org/docs/spec/client_server/r0.4.0.html#m-room-encrypted):

|    Field     |      Type      |     Description     |
|--------------|----------------|---------------------|

> | `signature`  | `string`       | Generated signature |

Reviewer: ```suggestion
| `signature`  | `string`       | Signature generated from `signed` using the private key from the same pair as the public key announced in `signed`'s `signatory_key` property. |
```

> | `signed`     | `SignedObject` | Object to be signed |

Reviewer: ```suggestion
| `signed`     | `SignedObject` | Object to be signed. |
```

Where:

* `SignedObject` is a map using the following structure:

| Parameter       |       Type       | Req. |                                                  Description                                                  |
|:----------------|:-----------------|:----:|:--------------------------------------------------------------------------------------------------------------|
| `sender`        | `string`         |  x   | User who sent the event. **Must** match the Matrix user ID of who sent the event.                             |
| `room_id`       | `string`         |  x   | Associated room ID. **Must** match the room ID where the event has been sent.                                 |
| `state_key`     | `string`         |  !   | State key of the event. Required if the event is a state event.                                               |
| `signatory`     | `string`         |      | The Matrix user ID who signed this data. If not present, defaults to the value defined in the `sender` field. |

> | `signatory_key` | `string`         |  x   | The public part of the key used for signing the event.                                                        |

Reviewer: ```suggestion
| `signatory_key` | `string`         |  x   | The public part of the key pair used for signing the event.                                                        |
```

| `algorithm`     | `string`         |  x   | Algorithm used for signing the content.                                                                       |
| `content`       | `object`         |  x   | The content of the event to be signed.                                                                        |

> 

Reviewer: Don't we also want to sign the event's type?

Author: yes !

Reviewer: Then you should add it :)

Reviewer: Nevermind, didn't notice it.

| `type`          | `string`         |  x   | Type of the Matrix event.                                                                                     |
| `state_key`     | `string`         |  !   | State key of the event. Required if the event is a state event.                                               |
| `signatory`     | `string`         |      | The Matrix user ID who signed this data. If not present, defaults to the value defined in the `sender` field. |
| `signatory_key` | `string`         |  x   | The public part of the key pair used for signing the event.                                                   |
| `algorithm`     | `string`         |  x   | Algorithm used for signing the content.                                                                       |
| `content`       | `object`         |  x   | The content of the event to be signed.                                                                        |

If the value of the `sender` field does not match the user ID sending the signed
Matrix event, or if the `room_id` does not match the room where the event is

> sent, then the signature **must** be considered as invalid.

Reviewer: Maybe make this a bit more generic like:

> If the value of any field referring to a property with the same name for the signed event doesn't match the value for the said property, then the signature **must** be considered as invalid.

Since the list of properties that need to match with the event that's being signed is getting quite long.
